bitcoind-rpcauth: script and service configuration

Stadicus opened the pull request:
This pull requests adds the recommended **rpcauth** methods for services to authenticate when using the bitcoind JSON RPC interface.

From the bitcoind help:
> `Username and HMAC-SHA-256 hashed password for JSON-RPC connections. The field <userpw> comes in the format: <USERNAME>:<SALT>$<HASH>. A canonical python script is included in share/rpcauth. The client then connects normally using the rpcuser=<USERNAME>/rpcpassword=<PASSWORD> pair of arguments. This option can be specified multiple times`

The rpcauth.py script from bitcoin repository to generate unique but persistent rpcauth credentials for bitcoind is added and extended to store the credentials directly in Redis.

To ensure unique credentials, the bitcoind-startpre script calls the command 'bbb-cmd.sh bitcoind refresh_rpcauth' if no valid rpcauth value is found in Redis (min length of 90 characters), or bitcoind:refresh-rpcauth is set to 1.

Configuration of bitcoind, lightningd, elects and bbbmiddleware is adjusted to use new authentication method:
* bitcoind.conf contains rpcauth setting: rpcuser plus salted/hashed password
* other services use rpcuser/rpcpassword to authenticate themselves

File access to config files containing rpcpassword is restricted to service user and group bitcoin. Two improvements should still be considered:
* storing the credentials in unsecured Redis db is not optimal
* Electrs should support rpcuser/rpcpassword without passing them through the command line (password visible in ps output)

As bitcoind does not manage funds, local rpcaccess (available only if system is already compromised) is not really critical and this solution good enough for an MVP.

The `redis.sh.inc` includes a new method `redis_require` while other methods like `redis_get` and `redis_set` don't fail the whole script if Redis is not available.

Stadicus commented on armbian/base/scripts/systemd-bitcoind-startpre.sh:
```diff
@@ -5,6 +5,26 @@
+RPCAUTH="$(redis_get 'bitcoind:rpcauth')"
+REFRESH_RPCAUTH="$(redis_get 'bitcoind:refresh-rpcauth')"
+
+if [ ${#RPCAUTH} -lt 90 ] || [ "${REFRESH_RPCAUTH}" -eq 1 ]; then
```
The `rpcauth` value usually is 103 characters long. I took 90 to have some error buffer, as the main check is that it's a real value and not the `xxx` set during build.

Stadicus commented on armbian/base/customize-armbian-rockpro64.sh:
```diff
@@ -207,7 +207,8 @@ else
 chown -R base:bitcoin /home/base/
-chmod -R 700 /home/base/.ssh/
+chmod 700 /home/base/.ssh/
+chmod 600 /home/base/.ssh/*
```
You are right. In that case it's not relevant as only the direct owner has access to the directory.

In other cases, the `x` permission is necessary for group members traversing the directory, but I also did not want to make all files executable.

E.g. if the `/etc/bitcoin` directory is set as 640 and owned by group `bitcoin`, the user `bitcoin` (which belongs to group `bitcoin`) cannot read the file `/etc/bitcoin/bitcoin.conf` despite having read access:

```
bitcoin@bitbox-base:/home/base$ ls -la /etc/bitcoin 
ls: cannot access '/etc/bitcoin/.': Permission denied
ls: cannot access '/etc/bitcoin/..': Permission denied
ls: cannot access '/etc/bitcoin/bitcoin.conf': Permission denied
total 0
d????????? ? ? ? ?            ? .
d????????? ? ? ? ?            ? ..
-????????? ? ? ? ?            ? bitcoin.conf
```

Will set this to recursive 600.

Stadicus commented on armbian/base/customize-armbian-rockpro64.sh:
```diff
@@ -563,7 +585,9 @@ fi
   mkdir -p /etc/bbbmiddleware/
+  chmod 750 /etc/bbbmiddleware/
   generateConfig "bbbmiddleware.conf.template" # --> /etc/bbbmiddleware/bbbmiddleware.conf
+  chmod 640 /etc/bbbmiddleware/*
```
There are not subdirectories, and if I would set them to to permissions without `x` (e.g. 640) even the owner would not be able to traverse them anymore. 

Maybe it's easier to just add and remove the permissions like this?
```
chmod -R u+rw,g+r,g-w,o-rw /etc/bbbmiddleware
```

0xB10C replied:
No, I don't think changes are needed if there are no subdirectories and there (probably )won't be any in the future. Similar goes for the other `chmod -R` comments.

Stadicus commented:
Simplified Makefile, removed debug comment and switched consistently to the following chmod syntax (this allows recursion without changing the `x` permission which is necessary to traverse directories)
```
chmod -R u+rw,g+r,g-w,o-rw /etc/bbbmiddleware
```
